concordium-node/src/bin/cli.rs

```diff
@@ -485,36 +487,104 @@ async fn maybe_do_out_of_band_catchup(
-                    }
-                } else {
-                    info!("Completed out of band catch-up from {}.", download_url);
-                }
-                // attempt to properly clean up the downloaded file.
-                if let Err(e) = path.close() {
-                    error!("Could not delete the downloaded file: {}", e);
-                }
-            }
-            Err(e) => {
-                // report error, but continue startup
-                error!("Downloading catch-up files failed: {}.", e)
+        info!("Starting out of band catch-up");
+        let genesis_block_hashes = regenesis_arc.blocks.read().unwrap().clone();
+        if let Err(e) =
+            import_missing_blocks(consensus, &genesis_block_hashes, download_url, data_dir_path)
+                .await
+        {
+            if import_stopped.load(atomic::Ordering::Acquire) {
+                info!("Out of band catchup stopped.");
+            } else {
+                info!("Could not complete out of band catch-up due to: {:#}.", e);
             }
+        } else {
+            info!("Completed out of band catch-up from {}.", download_url)
         }
     }
 }
 
-async fn download_catchup_files(
-    download_url: url::Url,
+// An index entry for a chunk of blocks. Its format must correspond to one
+// produced by `database-exporter`.
+#[derive(serde::Deserialize)]
+struct BlockChunkData {
+    // exported chunk of blocks' filename
+    filename:           String,
+    // genesis block index from which relative heights of blocks in the chunk are counted
+    genesis_index:      usize,
+    // relative height of the oldest block stored in the chunk
+    first_block_height: u64,
+    // relative height of the newest block stored in the chunk
+    last_block_height:  u64,
+}
+
+async fn import_missing_blocks(
+    consensus: &ConsensusContainer,
+    genesis_block_hashes: &[HashBytes],
+    index_url: &url::Url,
     data_dir_path: &Path,
-) -> anyhow::Result<TempPath> {
+) -> anyhow::Result<()> {
+    let current_genesis_index = genesis_block_hashes.len() - 1;
+    let last_finalized_block_height = consensus.get_last_finalized_block_height();
+
+    debug!("Current genesis index: {}", current_genesis_index);
+    debug!("Local last finalized block height: {}", last_finalized_block_height);
+
+    let mut index_reader = reqwest::get(index_url.clone())
+        .await
+        .context("Unable to download the catchup index file.")?
+        .bytes_stream()
+        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))
+        .into_async_read();
+    let mut first_line: String = String::new();
+    index_reader.read_line(&mut first_line).await?;
+
+    let index_genesis_block_hash = first_line
+        .strip_prefix("# genesis hash ")
+        .context(
+            "The catchup index file does not begin with a line containing the genesis block hash. \
+             Please contact the catchup service administrator.",
```

Comment:
I think it might be better to suggest checking that the URL is correct, and probably including the URL in the details.